[BUG]: Thrust/CUB should check that their `CUDA_ARCHITECTURES_RDC` list is non-empty before adding RDC targets.

### Is this a duplicate?

- [X] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Compile-time Error

### Component

CUB

### Describe the bug

(**Note:** Thrust/CUB historically conflated "relocatable device code" (RDC) and "CUDA dynamic parallelism" (CDP). References to architectures that do not support "RDC"  are actually referring to architectures that do not support CDPv1)

Certain architectures are filtered out by:
```cpp
list(FILTER CUB_CUDA_ARCHITECTURES_RDC EXCLUDE REGEX "53|62|72|90")
```
It seems that the RDC targets are still added, leading to the following errors:
```
CMake Error in cub/test/CMakeLists.txt:
  CUDA_ARCHITECTURES is empty for target
  "cub.cpp17.test.device_select_unique_by_key.cdp_1".


CMake Error in cub/test/CMakeLists.txt:
  CUDA_ARCHITECTURES is empty for target
  "cub.cpp17.test.device_select_unique_by_key.cdp_1".

...
```

### How to Reproduce

```bash
./ci/build_cub.sh "clang++" "17" "90"
```

### Expected behavior

CMake configuration works, no RDC targets are added. 

### Reproduction link

_No response_

### Operating System

_No response_

### nvidia-smi output

_No response_

### NVCC version

_No response_

--- a/cub/cmake/CubCudaConfig.cmake
+++ b/cub/cmake/CubCudaConfig.cmake
@@ -33,8 +33,15 @@ endif()
 
 option(CUB_FORCE_RDC "Enable separable compilation on all targets that support it." OFF)
 
+list(LENGTH CUB_CUDA_ARCHITECTURES_RDC rdc_arch_count)
+if (rdc_arch_count EQUAL 0)
+  message(NOTICE "Disabling CUB CDPv1 targets as no enabled architectures support it.")
+  set(CUB_ENABLE_RDC_TESTS OFF CACHE BOOL "" FORCE)
+  set(CUB_FORCE_RDC OFF CACHE BOOL "" FORCE)
+endif()
+
 #
-# Clang CUDA options 
+# Clang CUDA options
 #
 if ("Clang" STREQUAL "${CMAKE_CUDA_COMPILER_ID}")
   set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -Wno-unknown-cuda-version -Xclang=-fcuda-allow-variadic-functions")

--- a/thrust/cmake/ThrustCudaConfig.cmake
+++ b/thrust/cmake/ThrustCudaConfig.cmake
@@ -26,10 +26,16 @@ message(STATUS "THRUST_CUDA_ARCHITECTURES_RDC: ${THRUST_CUDA_ARCHITECTURES_RDC}"
 option(THRUST_ENABLE_RDC_TESTS "Enable tests that require separable compilation." ON)
 option(THRUST_FORCE_RDC "Enable separable compilation on all targets that support it." OFF)
 
+list(LENGTH THRUST_CUDA_ARCHITECTURES_RDC rdc_arch_count)
+if (rdc_arch_count EQUAL 0)
+  message(NOTICE "Disabling THRUST CDPv1 targets as no enabled architectures support it.")
+  set(THRUST_ENABLE_RDC_TESTS OFF CACHE BOOL "" FORCE)
+  set(THRUST_FORCE_RDC OFF CACHE BOOL "" FORCE)
+endif()
+
 #
 # Clang CUDA options
 #
 if ("Clang" STREQUAL "${CMAKE_CUDA_COMPILER_ID}")
   set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -Wno-unknown-cuda-version -Xclang=-fcuda-allow-variadic-functions")
 endif ()
-

